File: code/algo/getoffer/c++/4/resort_bybase.cpp
#include <iostream>
#include <queue>
#include <vector>

bool sequenceReconstruction(const std::vector<int> &nums,
                            const std::vector<std::vector<int>> &sequences) {
  int n = nums.size();
  std::vector<std::vector<int>> g(n + 1);
  std::vector<int> inDeg(n + 1, 0);

  for (const auto &sequence : sequences) {
    for (size_t i = 1; i < sequence.size(); ++i) {
      int x = sequence[i - 1];
      int y = sequence[i];
      g[x].push_back(y);
      inDeg[y]++;
    }
  }

  std::queue<int> q;
  for (int i = 1; i <= n; ++i) {
    if (inDeg[i] == 0) {
      q.push(i);
    }
  }

  while (!q.empty()) {
    if (q.size() > 1) {
      return false;
    }
    int x = q.front();
    q.pop();
    for (int y : g[x]) {
      if (--inDeg[y] == 0) {
        q.push(y);
      }
    }
  }

  return true;
}

int main() {
  std::vector<int> nums = {1, 2, 3};
  std::vector<std::vector<int>> sequences = {{1, 2}, {1, 3}};
  std::cout << "Can reconstruct sequence: "
            << (sequenceReconstruction(nums, sequences) ? "true" : "false")
            << std::endl;
  return 0;
}

File: code/algo/getoffer/go/4/resort_bybase.go
package main

func sequenceReconstruction(nums []int, sequences [][]int) bool {
	n := len(nums)
	g := make([][]int, n+1)
	inDeg := make([]int, n+1)
	for _, sequence := range sequences {
		for i := 1; i < len(sequence); i++ {
			x, y := sequence[i-1], sequence[i]
			g[x] = append(g[x], y)
			inDeg[y]++
		}
	}

	q := []int{}
	for i := 1; i <= n; i++ {
		if inDeg[i] == 0 {
			q = append(q, i)
		}
	}
	for len(q) > 0 {
		if len(q) > 1 {
			return false
		}
		x := q[0]
		q = q[1:]
		for _, y := range g[x] {
			if inDeg[y]--; inDeg[y] == 0 {
				q = append(q, y)
			}
		}
	}
	return true
}
